### ISSUE TITLE
Big quality difference

### ISSUE THREAD

**Issue by rkfg** [kijai/ComfyUI-PyramidFlowWrapper#28]
I compared the results produced by this wrapper with the original code, the wrapper generates a weird mess. I suspect something is wrong with the VAE or its settings. I used the code from the original repo and a slightly modified example prompt `A tsunami coming through the Red Square in Moscow, dynamic movement`.

Wrapper results:

https://github.com/user-attachments/assets/4120d0f4-c91a-4a70-adef-930aa3c86433

https://github.com/user-attachments/assets/f6e2fc43-e482-42a0-8c4f-1467f951e0fa

Quite weird, not smooth at all and random. Nothing like the original samples.

The HF repo code:

https://github.com/user-attachments/assets/43c76657-6882-400b-8aeb-f967a9009cfc

Miles ahead. I generated another video using the repo code but accidentally overwrote it with this one, it was a bit worse content-wise (the wave only appeared in the end) but still good visual quality-wise with people walking etc.

The wrapper uses much less VRAM, around 9-10 GB while the repo code takes up almost all of my 24 GB. But obviously I'd prefer that to get a good result. #18 could be related perhaps? Why I think it's a VAE issue: there's a bug with it that's easy to reproduce, it seems to lose/corrupt some internal state if you only change the VAE tile/window parameters after generation and re-run the workflow. CUI only recomputes the changed nodes and the result would look very distorted like in #25. If you change the parameters back, it would get even more corrupt and nothing like it was in the first run. So the only way to generate a video with a better quality is to restart CUI/change seed/change prompt so it generates from scratch again.

In general, the videos look low quality, perhaps even worse than what CogVideoX generates so something clearly isn't right.

Here are the parameters I used, mostly from the example, except I set the guidance to 9 as they suggest for 768p.
![2024-10-15_11-29-37](https://github.com/user-attachments/assets/2afb04ae-05f4-4d6b-a724-3fa63a3351fc)

EDIT: wait, there's more. The CUI produced videos are 7 seconds long while they should be 5. Perhaps that's the main problem?

**Comment by kijai** [comment]
The original code appends stuff to the prompts, there's always "cartoon style, worst quality, low quality, blurry, absolute black, absolute white, low res, extra limbs, extra digits, misplaced objects, mutated anatomy, monochrome, horror" in the negative if you leave it empty in their code, and there's always " ,hyper quality, Ultra HD, 8K" appended to the positive prompt.

With their example prompts my results have been fine, for example this one:

https://github.com/user-attachments/assets/1f2dc931-9ab4-4235-bb52-50b3702daea3

The VAE probably needs to be re-initialized when changing the settings, I noticed that too but keeping it at 256 has been fine, I think the original uses the 512 which doesn't really fit on consumer GPUs.

**Comment by kijai** [comment]
The FPS on the video combine node should be 24 to get the 5 seconds with temp 16, I just forgot to change that for the example workflows:

https://github.com/user-attachments/assets/ea5da537-e33f-4a9b-8ae4-d369737c6138

**Comment by rkfg** [comment]
> The original code appends stuff to the prompts, there's always "cartoon style, worst quality, low quality, blurry, absolute black, absolute white, low res, extra limbs, extra digits, misplaced objects, mutated anatomy, monochrome, horror" in the negative if you leave it empty in their code, and there's always " ,hyper quality, Ultra HD, 8K" appended to the positive prompt.

I used the python code from the README so nothing else was added (unless it's in the library itself as a default?):
```py
import torch
from PIL import Image
from pyramid_dit import PyramidDiTForVideoGeneration
from diffusers.utils import load_image, export_to_video

torch.cuda.set_device(0)
model_dtype, torch_dtype = 'bf16', torch.bfloat16   # Use bf16 (not support fp16 yet)

model = PyramidDiTForVideoGeneration(
    '../models/pyramidflow/pyramid-flow-sd3',                                         # The downloaded checkpoint dir
    model_dtype,
    model_variant='diffusion_transformer_768p',     # 'diffusion_transformer_384p'
)

model.vae.to("cuda")
model.dit.to("cuda")
model.text_encoder.to("cuda")
model.vae.enable_tiling()

prompt = "A tsunami coming through the Red Square in Moscow, dynamic movement"

with torch.no_grad(), torch.cuda.amp.autocast(enabled=True, dtype=torch_dtype):
    frames = model.generate(
        prompt=prompt,
        num_inference_steps=[20, 20, 20],
        video_num_inference_steps=[10, 10, 10],
        height=768,
        width=1280,
        temp=16,                    # temp=16: 5s, temp=31: 10s
        guidance_scale=9.0,         # The guidance for the first frame, set it to 7 for 384p variant
        video_guidance_scale=5.0,   # The guidance for the other video latent
        output_type="pil",
        save_memory=True,           # If you have enough GPU memory, set it to `False` to improve vae decoding speed
    )

export_to_video(frames, "./text_to_video_sample.mp4", fps=24)
```

In my examples above the CUI videos look oversaturated and the motion is choppy and random, while this script above always seems to produce smooth and more consistent motion. Another weirdness is that the script takes a lot of VRAM even during the generation phase, not only while decoding with VAE. Could be that VAE is already loaded at that point though, I have no idea. But it's a noticeable difference. Is it possible to compare the two using the exact same settings? I'm not sure how to set the seed here or there could be other differences.

**Comment by kijai** [comment]
> > The original code appends stuff to the prompts, there's always "cartoon style, worst quality, low quality, blurry, absolute black, absolute white, low res, extra limbs, extra digits, misplaced objects, mutated anatomy, monochrome, horror" in the negative if you leave it empty in their code, and there's always " ,hyper quality, Ultra HD, 8K" appended to the positive prompt.
> 
> I used the python code from the README so nothing else was added (unless it's in the library itself as a default?):
> 
> ```python
> import torch
> from PIL import Image
> from pyramid_dit import PyramidDiTForVideoGeneration
> from diffusers.utils import load_image, export_to_video
> 
> torch.cuda.set_device(0)
> model_dtype, torch_dtype = 'bf16', torch.bfloat16   # Use bf16 (not support fp16 yet)
> 
> model = PyramidDiTForVideoGeneration(
>     '../models/pyramidflow/pyramid-flow-sd3',                                         # The downloaded checkpoint dir
>     model_dtype,
>     model_variant='diffusion_transformer_768p',     # 'diffusion_transformer_384p'
> )
> 
> model.vae.to("cuda")
> model.dit.to("cuda")
> model.text_encoder.to("cuda")
> model.vae.enable_tiling()
> 
> prompt = "A tsunami coming through the Red Square in Moscow, dynamic movement"
> 
> with torch.no_grad(), torch.cuda.amp.autocast(enabled=True, dtype=torch_dtype):
>     frames = model.generate(
>         prompt=prompt,
>         num_inference_steps=[20, 20, 20],
>         video_num_inference_steps=[10, 10, 10],
>         height=768,
>         width=1280,
>         temp=16,                    # temp=16: 5s, temp=31: 10s
>         guidance_scale=9.0,         # The guidance for the first frame, set it to 7 for 384p variant
>         video_guidance_scale=5.0,   # The guidance for the other video latent
>         output_type="pil",
>         save_memory=True,           # If you have enough GPU memory, set it to `False` to improve vae decoding speed
>     )
> 
> export_to_video(frames, "./text_to_video_sample.mp4", fps=24)
> ```
> 
> In my examples above the CUI videos look oversaturated and the motion is choppy and random, while this script above always seems to produce smooth and more consistent motion. Another weirdness is that the script takes a lot of VRAM even during the generation phase, not only while decoding with VAE. Could be that VAE is already loaded at that point though, I have no idea. But it's a noticeable difference. Is it possible to compare the two using the exact same settings? I'm not sure how to set the seed here or there could be other differences.

Yes the pipeline itself adds the prompts, it felt silly so I didn't include them, though they are in my example workflow by default.

I did a lot of memory optimizations, mostly just offloading to RAM though, nothing that would affect quality if you stick with bf16.

**Comment by rkfg** [comment]
Very interesting that they do that! It indeed looks silly but if it was in the training data maybe it makes sense to add (not automatically of course). I corrected the prompts and now the output is just as good, thank you! Wouldn't hurt to mention this in the README somewhere because otherwise the result looks too bad by default.

**Comment by rkfg** [comment]
>I didn't include them, though they are in my example workflow by default.

In the text2vid example there's no negative prompt though, only the positive suffix. And the negative prompt makes a whole ton of difference:

Without:

https://github.com/user-attachments/assets/b5c9e25f-a295-4cbe-aff4-1cc2866fc08f

With:

https://github.com/user-attachments/assets/b1381490-bbf8-4593-9f76-e57195353b56

**Comment by kijai** [comment]
Right, good point, I'll add that.
